core/scripts/i18n/collect-strings.js

```diff
@@ -613,43 +614,100 @@ function writeStringsToCtcFiles(locale, strings) {
+ *   Either `placeholders` should be the same, or `message` or `description`
+ *   should be changed to explain the need for different `placeholders`.
  *
+ * Modifies `strings` in place. Returns list of colliding messages that were fixed.
+ * @see https://github.com/GoogleChrome/lighthouse/blob/main/core/lib/i18n/README.md#collisions
  * @param {Record<string, CtcMessage>} strings
  */
 function resolveMessageCollisions(strings) {
-  /** @type {Map<string, Array<CtcMessage>>} */
-  const stringsByMessage = new Map();
-
-  // Group all the strings by their message.
-  for (const ctc of Object.values(strings)) {
-    const collisions = stringsByMessage.get(ctc.message) || [];
-    collisions.push(ctc);
-    stringsByMessage.set(ctc.message, collisions);
-  }
+  const entries = Object.entries(strings).map(([key, ctc]) => ({key, ctc}));
+
+  const stringsByMessage = arrayGroupToMap(entries, (entry) => entry.ctc.message);
 
   /** @type {Array<CtcMessage>} */
-  const allCollisions = [];
+  const fixedCollisions = [];
   for (const messageGroup of stringsByMessage.values()) {
     // If this message didn't collide with anything else, skip it.
     if (messageGroup.length <= 1) continue;
 
-    // If group shares both message and description, they can be translated as if a single string.
-    const descriptions = new Set(messageGroup.map(ctc => ctc.description));
-    if (descriptions.size <= 1) continue;
+    // For each group of matching message+description, placeholders must also match.
+    const stringsByDescription = arrayGroupToMap(messageGroup, (entry) => entry.ctc.description);
+    for (const descriptionGroup of stringsByDescription.values()) {
+      if (!doPlaceholdersMatch(descriptionGroup)) {
+        throw new Error('string collision: `message` and `description` are the same but differ in `placeholders`:\n' +
+          descriptionGroup.map(entry => `key: ${entry.key}\n`).join('') +
+          'make `placeholders` match or update `message` or `description`s to explain why they don\'t match');
+      }
+    }
+
+    // If the entire message group has the same description and placeholders,
+    // they can be translated as if a single message, no meaning needed.
+    if (stringsByDescription.size <= 1) continue;
 
     // We have duplicate messages with different descriptions. Disambiguate using `meaning` for TC.
-    for (const ctc of messageGroup) {
+    for (const {ctc} of messageGroup) {
       ctc.meaning = ctc.description;
     }
-    allCollisions.push(...messageGroup);
+    fixedCollisions.push(...messageGroup.map(({ctc}) => ctc));
   }
 
-  // Check that the known collisions match our known list.
-  const collidingMessages = allCollisions.map(collision => collision.message).sort();
+  return fixedCollisions;
+}
+
+/**
+ * Check that the known collisions match our known list.
+ * @param {Array<CtcMessage>} fixedCollisions
+ */
+function checkKnownCollisions(fixedCollisions) {
```

Comment:
split this into a different function because otherwise the only way to test `resolveMessageCollisions` would be against all the strings collected from lighthouse (since all the known collisions are asserted to be there). If this file gets much more testing, it probably just needs to not interleave reading/writing disk and the parsing/string creation code